### DISCUSSION TITLE
Is it possible to kill the process it started when gsudo is terminated?

### DISCUSSION THREAD

**Discussion by Glavo** [gerardog/gsudo#186] [Q&A]
@gerardog I am starting another subprocess in my program. To get the subprocess to run with admin privileges, I started it with gsudo.

But I now have a problem: I need to terminate this child process in my program, but when I use gsudo, `TerminateProcess` only terminates gsudo itself, and the process started by gsudo is still running.

So, how should I terminate this process?

**Comment by gerardog** [comment]
Interesting question.
What language are you using?

Two options here, 
1. Using [Windows Job Objects](https://learn.microsoft.com/en-us/windows/win32/procthread/job-objects)
   A Job allows to group several processes as a unit. If gsudo is groupped with the elevated process, if you kill gsudo that would kill the child process. The problem would be, who creates the job? probably gsudo should. Probably other processes won't be able to create the job without elevating first.

2. Manual killing
  You need to get the PID of the elevated process (the one launched by gsudo), and then kill it... but killing it also requires elevation...

For example in PowerShell
![image](https://user-images.githubusercontent.com/3901474/193873165-6fcaccdb-9f5b-437c-9e68-0a9c0b98edb2.png)

In a nutshell
   - Launch gsudo, get gsudo PID.
   - Enumerate gsudo child processes, find the interesting one (or all)
   - Elevate to kill it.

**Reply by Glavo** [reply]
@gerardog 

I am using Java.

Well, neither solution is so easy for my needs. I don't want to give admin rights to programs that don't have digital signatures (sadly, I don't have a digital certificate, so can't write a wrapper myself to solve the problem), I also only want to request permission from the user once. In the end I also need to get the stdout of the process and wait for it to exit (so I used gsudo).

I'm curious about the first solution, is it possible to make gsudo provide an option to use it?

**Comment by gerardog** [comment]
I created an issue to track this change.  #193

**Comment by bilbothebaggins** [comment]
@Glavo If you need to Kill/Terminate arbitrary child processes, IMO you always should kill the child process tree. (That is, do it like Process Explorer and enumerate all child processes.) Otherwise you're always in the situation that if anything in the child spawns another process (powershell Start-Job anyone) than you won't catch it.

Btw. @gerardog ....

> Manual killing
> ... and then kill it... but killing it also requires elevation...

Killing an elevated process in the same logon session [does not require elevation](https://stackoverflow.com/questions/75243609/how-does-pskill-work-across-the-uac-elevation-boundary).

`OpenProcess(SYNCHRONIZE | PROCESS_QUERY_LIMITED_INFORMATION | PROCESS_TERMINATE)` work even from a non elevated parent process to an elevated child process, at least in the same-user case.
